x-pack/modules/azure/lib/filters/azure_event.rb

```diff
@@ -0,0 +1,246 @@
+
+require "logstash/filters/base"
+require "logstash/namespace"
+require "logstash/json"
+
+module LogStash
+  module Filters
+    class AzureEvent < LogStash::Filters::Base
+
+      config_name "azure_event"
+
+      # When all heuristics fail to identify a category, this category will be used if and only if the default_match fields are present, and the resource has been parsed correctly
+      config :default_category, :validate => :string, :default => "Administrative"
+
+      # These fields must be present to match to the default default category
+      config :default_match, :validate => :array, :default => ["operationName", "level"]
+
+      # Possible [azure][category] values for [azure][group] = "activity_log", the category names must be mutually exclusive between groups
+      config :activity_log_categories, :validate => :array, :default => ["Administrative", "ServiceHealth", "Alert", "Autoscale", "Security"]
+
+      # Possible [azure][category] values for [azure][group] = "sql_diagnostics", the category names must be mutually exclusive between groups
+      config :sql_diagnostics_categories, :validate => :array, :default =>
+          ["Metric", "Blocks", "Errors", "Timeouts", "QueryStoreRuntimeStatistics", "QueryStoreWaitStatistics", "DatabaseWaitStatistics", "SQLInsights"]
+
+      # Append values to the `tags` field when there has been not been a match to a known azure event group or category.
+      config :tag_on_failure, :validate => :array, :default => ["_azure_event_failure"]
+
+      ACTIVITY_LOG_GROUP = "activity_log".freeze
+      SQL_DIAGNOSTICS_GROUP = "sql_diagnostics".freeze
+
+      def register
+        # do nothing
+      end
+
+      def filter(event)
+        resource_id = event.get("resourceId")
+        if resource_id && add_resources(resource_id, event) && add_group_and_category(event)
+          # rename properties to include the group/category to help avoid mapping errors
+          if event.get("properties")
+            group = event.get("[azure][group]")
+            category = event.get("[azure][category]")
+            explode_properties(group, category, event)
+            event.set("[#{group}_#{category}_properties]", event.get("properties"))
+            event.remove("properties")
+          end
+          filter_matched(event)
+        else
+          @tag_on_failure.each do |tag|
+            event.tag(tag)
+          end
+          logger.trace("Failed to parse event as an Azure Event", :event => event) if logger.trace?
+        end
+      end
+
+      # public methods may change state of event, and should return true if successful
+
+      ##
+      # Parses the resourceId to pull out the relevant parts into the event
+      # return true if the common set of fields can be extracted from the resourceId, false otherwise
+      def add_resources(resource_id, event)
+        common_resources = explode_common_resources(resource_id)
+        if common_resources
```

Comment:
more of an overall code design suggestion, keeping the indentation level low is typically preferred. as it makes reading the code easier.

For example, here we can do 

```ruby
     def add_resources(resource_id, event)
       common_resources = explode_common_resources(resource_id)
       return false if common_resources.nil?
       # everything else, now with 1 less indentation level 
```

This also makes it clearer very early in the method that we short circuit the rest of the logic in a particular situation.

Reply:
I haven't ever heard of keeping indentation low is a goal for readability.  

Rather, we should discourage multiple exit points from a method...which implies if there are a lot of conditionals then there are a lot of indentation to keep the flow to the single exit condition.

Reply:
hmm.. googling on this a bit , and it seems that we are both right and wrong.  The arguments alot like tabs vs. spaces (except with tabs vs. spaces there is a clear winner ;) ).

Reply:
Here's where my research on indentation led me

![Fireball!](https://i.imgur.com/94kLRVp.jpg)

Reply:
> you have the mental weight of "what's in the upcoming elses?" when looking at the code.

well put, I think this is my main concern. I tend to read code in a breadth first manner to gauge the overall flow, and reaching a conditional (or a set of nested conditionals) always makes me push into my brain stack:
stack << "remember to check what the else for this if does"
stack << "remember to check what the else for this if does"
stack << "remember to check what the else for this if does"

I can be persuaded otherwise though :)

> we should discourage multiple exit points from a method...which implies if there are a lot of conditionals then there are a lot of indentation to keep the flow to the single exit condition.

even if is one exit (which thankfully your code is really well done in that way), I still see a benefit of making the only "abort" explicit early in the method.